rest/bytes_read_public_api_test.go

```diff
@@ -0,0 +1,374 @@
+	rt := NewRestTester(t, &RestTesterConfig{SyncFn: channels.DocChannelsSyncFunction})
+	defer rt.Close()
+
+	// create a user to authenticate as for public api calls and assert the stat hasn't incremented as a result
+	rt.CreateUser("greg", []string{"ABC"})
+	_, ok := base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, 0)
+	require.True(t, ok)
+
+	// use public api to put a doc through SGW then assert the stat has increased
+	input := `{"foo":"bar", "channels":["ABC"]}`
+	inputBytes := []byte(input)
+	resp := rt.SendUserRequest(http.MethodPut, "/{{.keyspace}}/doc1", input, "greg")
+	RequireStatus(t, resp, http.StatusCreated)
+
+	_, ok = base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(len(inputBytes)))
+	require.True(t, ok)
+
+	// send admin request assert that the public rest count doesn't increase
+	resp = rt.SendAdminRequest(http.MethodGet, "/{{.keyspace}}/doc1", "")
+	RequireStatus(t, resp, http.StatusOK)
+	_, ok = base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(len(inputBytes)))
+	require.True(t, ok)
+
+	// send user request that has empty body, asser the stat doesn't increase
+	resp = rt.SendUserRequest(http.MethodGet, "/{{.keyspace}}/doc1", "", "greg")
+	RequireStatus(t, resp, http.StatusOK)
+	_, ok = base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(len(inputBytes)))
+	require.True(t, ok)
+
+	// assert blipsync connection doesn't increment stat
+	resp = rt.SendUserRequest(http.MethodGet, "/{{.db}}/_blipsync", "", "greg")
+	RequireStatus(t, resp, http.StatusUpgradeRequired)
+	_, ok = base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(len(inputBytes)))
+	require.True(t, ok)
+
+	srv := httptest.NewServer(rt.TestMetricsHandler())
+	defer srv.Close()
+	httpClient := http.DefaultClient
+
+	// test metrics endpoint, assert the stat doesn't increment
+	response, err := httpClient.Get(srv.URL + "/_metrics")
+	require.NoError(t, err)
+	assert.Equal(t, http.StatusOK, response.StatusCode)
+
+	_, ok = base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(len(inputBytes)))
+	require.True(t, ok)
+
+	// test public endpoint but one that doesn't access a db and assert that doesn't increment stat
+	resp = rt.SendUserRequest(http.MethodGet, "/", "", "greg")
+	RequireStatus(t, resp, http.StatusOK)
+	// assert the stat doesn't increment
+	_, ok = base.WaitForStat(func() int64 {
+		return rt.GetDatabase().DbStats.DatabaseStats.PublicRestBytesRead.Value()
+	}, int64(len(inputBytes)))
+	require.True(t, ok)
+
+	// send another public request (this time POST) to check stat increments, but check it increments by correct bytes value
```

Comment:
add a test here to make sure that a stat increases for a user who doesn't have access to the db

Reply:
This is done inside TestBytesReadAuthFailed